### PR TITLE
Add all vector opcodes (except related to masks and shuffles)

### PR DESCRIPTION
**Pull request by gita-omr**
- removed all old vector opcodes
- added vector operations with two types (e.g. vconv and vcast)
- added vector operations that return scalar (e.g. vfirstNonZero and reductions in the future)
- replaced isVector() with isVectorResult()
- as a result, current set of vector operations: vabs vadd vand vcall vcalli vcmp* vdiv vfma vindexVector vload vloadi vmax vmin vmul vneg vnot vor vorUnchecked vreturn vRegLoad vRegStore vbitselect vsetelem vsplats vsqrt vstore vstorei vsub vxor vfirstNonZero vgetelem vcast vconv

### CODE REVIEW

**Comment by gita-omr** [comment]
Jenkins build all

**Comment by gita-omr** [comment]
Expected conflicts with https://github.com/eclipse/omr/pull/6495. Will fix soon.

**Comment by gita-omr** [comment]
More conflict resolution.

**Comment by gita-omr** [comment]
Fixed x86-64 build

**Comment by gita-omr** [comment]
Fixed yet another typo during conflict resolution.

**Comment by gita-omr** [comment]
Fixed yet more typos during conflict resolution...

**Comment by gita-omr** [comment]
MAC failure might need some rework.  So making WIP for now.

**Comment by gita-omr** [comment]
Fixed line ending

**Comment by gita-omr** [comment]
Fixed Copyright.

**Comment by gita-omr** [comment]
Rebased.

**Comment by gita-omr** [comment]
Fixed JitBuilder failure.

**Comment by gita-omr** [comment]
The MacOS and PPCLE failures seem to be unrelated...

**Comment by gita-omr** [comment]
Rebased

**Comment by gita-omr** [comment]
Corrected latest commit (reduction opcodes)

**Comment by gita-omr** [comment]
Jenkins build all

**Comment by gita-omr** [comment]
Fixed line ending in the last commit.

**Comment by gita-omr** [comment]
Jenkins build all

**Comment by gita-omr** [comment]
Made printing of vector opcodes more general to accommodate reduction and conversion opcodes.

File: compiler/x/codegen/UnaryEvaluator.cpp [hunk 1]
```diff
@@ -45,9 +45,9 @@ TR::Register *OMR::X86::TreeEvaluator::unaryVectorArithmeticEvaluator(TR::Node *
    TR::InstOpCode::Mnemonic regRegOpcode;
    TR::InstOpCode::Mnemonic regMemOpcode;
 
-   switch (node->getOpCodeValue())
+   switch (node->getOpCode().getVectorOperation())
       {
-      case TR::vdsqrt:
+      case TR::vsqrt:
          regRegOpcode = supportsAvx ? OMR::InstOpCode::VSQRTPDRegReg : OMR::InstOpCode::SQRTPDRegReg;
```
Comment (BradleyWood) [inline compiler/x/codegen/UnaryEvaluator.cpp:51]: Assert that the element type is a double for now. Other types can be added later.

**Review by BradleyWood — COMMENTED** [review]
x86 changes look good

File: compiler/x/codegen/UnaryEvaluator.cpp (hunk 1 above)
Comment (gita-omr) [reply compiler/x/codegen/UnaryEvaluator.cpp:55]: Added an ASSERT.

**Comment by gita-omr** [comment]
Addressed comments.

File: compiler/il/VectorOperations.enum [hunk 2]
```diff
@@ -18,308 +18,758 @@
...
+VECTOR_OPERATION_MACRO(\
+   /* .operation               = */ vstorei, \
+   /* .name                    = */ "vstorei", \
+   /* .properties1             = */ ILProp1::Store | ILProp1::Indirect | ILProp1::TreeTop | ILProp1::HasSymbolRef, \
+   /* .properties2             = */ ILProp2::MayUseSystemStack, \
+   /* .properties3             = */ ILProp3::LikeDef, \
+   /* .properties4             = */ 0, \
+   /* .dataType                = */ TR::NoType, \
+   /* .typeProperties          = */ ILTypeProp::VectorResult, \
+   /* .childProperties         = */ TWO_SAME_CHILD(ILChildProp::UnspecifiedChildType), \
```
Comment (0xdaryl) [inline compiler/il/VectorOperations.enum:670]: Should this be:
```
TWO_CHILD(TR::Address, ILChildProp::UnspecifiedChildType), \
```

File: compiler/il/VectorOperations.enum [hunk 3]
```diff
@@ -18,308 +18,758 @@
...
+VECTOR_OPERATION_MACRO(\
+   /* .operation               = */ vloadi, \
+   /* .name                    = */ "vloadi", \
+   /* .properties1             = */ ILProp1::LoadVar | ILProp1::Indirect | ILProp1::HasSymbolRef, \
+   /* .properties2             = */ ILProp2::ValueNumberShare | ILProp2::SupportedForPRE| ILProp2::MayUseSystemStack, \
+   /* .properties3             = */ ILProp3::LikeUse, \
+   /* .properties4             = */ 0, \
+   /* .dataType                = */ TR::NoType, \
+   /* .typeProperties          = */ ILTypeProp::VectorResult, \
+   /* .childProperties         = */ ONE_CHILD(ILChildProp::UnspecifiedChildType), \
```
Comment (0xdaryl) [inline compiler/il/VectorOperations.enum:270]: Should this be:
```
ONE_CHILD(TR::Address), \
```

File: compiler/il/VectorOperations.enum [hunk 4]
```diff
@@ -18,308 +18,758 @@
...
+   /* .properties3             = */ 0, \
+   /* .properties4             = */ 0, \
+   /* .dataType                = */ TR::NoType, \
+   /* .typeProperties          = */ ILTypeProp::VectorResult, \
+   /* .childProperties         = */ TWO_SAME_CHILD(ILChildProp::UnspecifiedChildType), \
+   /* .swapChildrenOperation   = */ TR::vBadOperation, \
+   /* .reverseBranchOperation  = */ TR::vBadOperation, \
+   /* .booleanCompareOpCode    = */ TR::BadILOp, \
+   /* .ifCompareOpCode         = */ TR::BadILOp, \
+   /* .description             =    vector logical OR */ \
```
Comment (0xdaryl) [inline compiler/il/VectorOperations.enum:387]: Can this description be updated to distinguish this opcode from "vor"?  i.e., what's the difference?

File: compiler/il/VectorOperations.enum [hunk 5]
```diff
@@ -18,308 +18,758 @@
...
+   /* .properties3             = */ 0, \
+   /* .properties4             = */ 0, \
+   /* .dataType                = */ TR::Int32, \
+   /* .typeProperties          = */ TR::NoType, \
+   /* .childProperties         = */ TWO_SAME_CHILD(ILChildProp::UnspecifiedChildType), \
+   /* .swapChildrenOperation   = */ TR::vBadOperation, \
+   /* .reverseBranchOperation  = */ TR::vBadOperation, \
+   /* .booleanCompareOpCode    = */ TR::BadILOp, \
+   /* .ifCompareOpCode         = */ TR::BadILOp, \
+   /* .description             =    get vector element */ \
```
Comment (0xdaryl) [inline compiler/il/VectorOperations.enum:723]: Is this the correct description?

File: compiler/il/OMRILOps.hpp [hunk 6]
```diff
@@ -117,21 +87,53 @@ class ILOpCode
...
+                             operation * TR::NumVectorTypes * TR::NumVectorTypes +
+                             (srcVectorType - TR::NumScalarTypes) * TR::NumVectorTypes +
+                             (resVectorType - TR::NumScalarTypes));
+      }
+
+   bool isTwoTypeVectorOpCode()
+         {
+         if (!isVectorOpCode()) return false;
+
+         return (getVectorOperation() >= TR::firstTwoVectorTypeOperation);
```
Comment (0xdaryl) [inline compiler/il/OMRILOps.hpp:134]: Can you make the naming consistent with the thing you're checking?  i.e., is it "TwoTypeVector" or "TwoVectorType"

**Review by 0xdaryl — CHANGES_REQUESTED** [review]

File: compiler/il/OMRILOps.cpp [hunk 7]
```diff
@@ -305,3 +303,62 @@ OMR::ILOpCode::getCompareType(TR::ILOpCodes op)
...
+   {
+   if (t1.isVector() && t2.isVector())
+      return TR::ILOpCode::createVectorOpCode(TR::vcast, t1, t2);
+
+   if (t1.isVector() || t2.isVector())
+      return TR::BadILOp;
+
+
+   TR_ASSERT(t1 < TR::NumOMRTypes, "conversion opcode from unexpected datatype %s requested", t1.toString());
+   TR_ASSERT(t2 < TR::NumOMRTypes, "conversion opcode to unexpected datatype %s requested", t2.toString());
```
[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/il/OMRILOps.cpp:342]: This method existed before, just in another class. I a bit worry about making it FATAL as a part of this PR...

File: compiler/il/OMRILOps.cpp [hunk 8]
```diff
@@ -305,3 +303,62 @@ OMR::ILOpCode::getCompareType(TR::ILOpCodes op)
...
+   }
+
+TR::ILOpCodes
+OMR::ILOpCode::getDataTypeBitConversion(TR::DataType t1, TR::DataType t2)
+   {
+   if (t1.isVector() || t2.isVector())
+      return TR::BadILOp;
+
+   TR_ASSERT(t1 < TR::NumOMRTypes, "conversion opcode from unexpected datatype %s requested", t1.toString());
+   TR_ASSERT(t2 < TR::NumOMRTypes, "conversion opcode to unexpected datatype %s requested", t2.toString());
```
[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/il/OMRILOps.cpp:353]: The same as above.

**Comment by gita-omr** [comment]
Addressed comments from @0xdaryl

File: compiler/aarch64/codegen/OMRTreeEvaluator.cpp [hunk 9]
```diff
@@ -3247,6 +3068,30 @@ OMR::ARM64::TreeEvaluator::vstoreEvaluator(TR::Node *node, TR::CodeGenerator *cg
    return commonStoreEvaluator(node, TR::InstOpCode::vstrimmq, cg);
    }
 
+TR::Register *
+OMR::ARM64::TreeEvaluator::vindexVectorEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   return OMR::ARM64::TreeEvaluator::unImpOpEvaluator(node, cg);
```
[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/aarch64/codegen/OMRTreeEvaluator.cpp]: Done. Thanks!

File: compiler/il/OMRILOpCodeProperties.hpp [hunk 10]
```diff
@@ -83,13 +83,13 @@
    dataType, \
    typeProps, \
    childProps, \
-   swapChildrenOpcode, \
-   reverseBranchOpcode, \
+   swapVectorOperation, \
+   reverseVectorOperation, \
```
[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/il/OMRILOpCodeProperties.hpp]: Done

File: compiler/p/codegen/FPTreeEvaluator.cpp [hunk 11]
```diff
@@ -1732,13 +1679,13 @@ TR::Register *OMR::Power::TreeEvaluator::vRegLoadEvaluator(TR::Node *node, TR::C
...
    if (globalReg == NULL)
       {
-      if (node->getOpCode().getOpCodeValue() == TR::vbRegLoad ||
-          node->getOpCode().getOpCodeValue() == TR::vsRegLoad ||
-          node->getOpCode().getOpCodeValue() == TR::viRegLoad ||
-          node->getOpCode().getOpCodeValue() == TR::vlRegLoad)
+      if (node->getOpCode().getVectorResultDataType().getVectorElementType() == TR::Int8 ||
+          node->getOpCode().getVectorResultDataType().getVectorElementType() == TR::Int16 ||
+          node->getOpCode().getVectorResultDataType().getVectorElementType() == TR::Int32 ||
+          node->getOpCode().getVectorResultDataType().getVectorElementType() == TR::Int64)
```
[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/p/codegen/FPTreeEvaluator.cpp]: Done

File: compiler/x/codegen/OMRCodeGenerator.cpp [hunk 12]
```diff
@@ -1077,11 +1046,36 @@ OMR::X86::CodeGenerator::getSupportsOpCodeForAutoSIMD(TR::ILOpCode opcode, TR::D
...
+      case TR::vsplats:
          if (et == TR::Int32 || et == TR::Int64 || et == TR::Float || et == TR::Double)
             return ot.getVectorLength() == TR::VectorLength128;
          else
             return false;
+      case TR::vsqrt:
+          if (et == TR::Double)
+              return true;
+          else
+              return false;
```
[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/x/codegen/OMRCodeGenerator.cpp]: Done

File: compiler/z/codegen/OMRCodeGenerator.cpp [hunk 13]
```diff
@@ -4719,72 +4719,67 @@ bool OMR::Z::CodeGenerator::isDispInRange(int64_t disp)
...
+   TR_ASSERT_FATAL(opcode.isVectorOpCode(), "getSupportsOpCodeForAutoSIMD expects vector opcode\n");
 
    TR::DataType ot = opcode.getVectorResultDataType();
 
    if (ot.getVectorLength() != TR::VectorLength128) return false;
 
    TR::DataType et = opcode.getVectorResultDataType().getVectorElementType();
 
-   TR_ASSERT_FATAL(et == TR::Int8 || et == TR::Int16 || et == TR::Int32 || et == TR::Int64 || et == TR::Float || et == TR::Double, "DataType %s is not supported for vectorization", et.toString());
+  TR_ASSERT_FATAL(et == TR::Int8 || et == TR::Int16 || et == TR::Int32 || et == TR::Int64 || et == TR::Float || et == TR::Double,
```
[the remark this answers is not in the document]
Comment (gita-omr) [reply compiler/z/codegen/OMRCodeGenerator.cpp]: Done

File: compiler/il/VectorOperations.enum (hunk 3 above)
Comment (gita-omr) [reply compiler/il/VectorOperations.enum]: Done

File: compiler/il/VectorOperations.enum (hunk 4 above)
Comment (gita-omr) [reply compiler/il/VectorOperations.enum]: Done

File: compiler/il/VectorOperations.enum (hunk 5 above)
Comment (gita-omr) [reply compiler/il/VectorOperations.enum]: Done

File: compiler/il/OMRILOps.hpp (hunk 6 above)
Comment (gita-omr) [reply compiler/il/OMRILOps.hpp]: Done: TwoTypeVector

**Comment by gita-omr** [comment]
Fixed line ending

**Comment by 0xdaryl** [comment]
Jenkins build all

File: compiler/il/VectorOperations.enum (hunk 2 above)
Comment (0xdaryl) [reply compiler/il/VectorOperations.enum:670]: All looks good except this comment is outstanding.

**Comment by gita-omr** [comment]
Fixed vstorei child proterties

**Comment by 0xdaryl** [comment]
Jenkins build all

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**